File: mcp_delegation_server/benchmark/scorers.py
"""Pluggable scorers for benchmarking LLM outputs."""

from __future__ import annotations

import json
import re
from abc import ABC, abstractmethod
from dataclasses import asdict, dataclass, field
from typing import Any

from mcp_delegation_server.runners.base import BaseRunner, RunnerResult


@dataclass
class ScoreResult:
    """Evaluation score and metrics returned by a scorer."""

    scorer_name: str
    score: float  # Normalized score (e.g. 0.0 to 100.0) or raw metric value
    unit: str = "points"  # "points", "ms", "usd", "percent", "tokens/sec"
    details: dict[str, Any] = field(default_factory=dict)
    reasoning: str | None = None

    def to_dict(self) -> dict[str, Any]:
        return asdict(self)


class BaseScorer(ABC):
    """Abstract base class for benchmark scorers."""

    @property
    @abstractmethod
    def name(self) -> str:
        pass

    @abstractmethod
    async def score(
        self,
        task: str,
        result: RunnerResult,
        reference_answer: str | None = None,
        all_results: list[RunnerResult] | None = None,
    ) -> ScoreResult:
        """Compute an evaluation score for a runner's result."""
        pass


class LatencyScorer(BaseScorer):
    """Scores runners based on response latency and generation throughput."""

    @property
    def name(self) -> str:
        return "latency"

    async def score(
        self,
        task: str,
        result: RunnerResult,
        reference_answer: str | None = None,
        all_results: list[RunnerResult] | None = None,
    ) -> ScoreResult:
        if not result.success:
            return ScoreResult(
                scorer_name=self.name,
                score=0.0,
                unit="ms",
                details={"latency_ms": result.latency_ms, "tokens_per_second": 0.0},
                reasoning="Execution failed",
            )

        tokens_per_sec = 0.0
        if result.latency_ms > 0 and result.output_tokens > 0:
            tokens_per_sec = round(result.output_tokens / (result.latency_ms / 1000.0), 2)

        return ScoreResult(
            scorer_name=self.name,
            score=result.latency_ms,
            unit="ms",
            details={
                "latency_ms": result.latency_ms,
                "output_tokens": result.output_tokens,
                "tokens_per_second": tokens_per_sec,
            },
            reasoning=f"Completed in {result.latency_ms} ms (~{tokens_per_sec} tokens/sec)",
        )


class CostScorer(BaseScorer):
    """Scores runners based on financial token cost in USD."""

    @property
    def name(self) -> str:
        return "cost"

    async def score(
        self,
        task: str,
        result: RunnerResult,
        reference_answer: str | None = None,
        all_results: list[RunnerResult] | None = None,
    ) -> ScoreResult:
        total_tokens = result.input_tokens + result.output_tokens
        return ScoreResult(
            scorer_name=self.name,
            score=result.estimated_cost_usd,
            unit="usd",
            details={
                "input_tokens": result.input_tokens,
                "output_tokens": result.output_tokens,
                "total_tokens": total_tokens,
                "estimated_cost_usd": result.estimated_cost_usd,
            },
            reasoning=f"Cost: ${result.estimated_cost_usd:.6f} across {total_tokens} total tokens",
        )


class ExactMatchScorer(BaseScorer):
    """Scores outputs against a reference answer using exact match and token overlap."""

    @property
    def name(self) -> str:
        return "reference_match"

    async def score(
        self,
        task: str,
        result: RunnerResult,
        reference_answer: str | None = None,
        all_results: list[RunnerResult] | None = None,
    ) -> ScoreResult:
        if not reference_answer:
            return ScoreResult(
                scorer_name=self.name,
                score=0.0,
                unit="percent",
                details={"evaluated": False},
                reasoning="No reference answer provided for matching",
            )

        if not result.success or not result.output:
            return ScoreResult(
                scorer_name=self.name,
                score=0.0,
                unit="percent",
                details={"exact_match": False, "token_overlap": 0.0},
                reasoning="Runner did not produce valid output",
            )

        target = reference_answer.strip()
        actual = result.output.strip()

        # Exact match
        if actual == target:
            return ScoreResult(
                scorer_name=self.name,
                score=100.0,
                unit="percent",
                details={"exact_match": True, "token_overlap": 1.0},
                reasoning="Exact verbatim match with reference answer",
            )

        # Case-insensitive stripped match
        if actual.lower() == target.lower():
            return ScoreResult(
                scorer_name=self.name,
                score=95.0,
                unit="percent",
                details={"exact_match": False, "case_insensitive_match": True},
                reasoning="Case-insensitive match with reference answer",
            )

        # Substring match
        if target.lower() in actual.lower():
            return ScoreResult(
                scorer_name=self.name,
                score=80.0,
                unit="percent",
                details={"substring_match": True},
                reasoning="Reference answer is fully contained in output",
            )

        # Word token overlap (Jaccard / F1 similarity)
        target_words = set(re.findall(r"\w+", target.lower()))
        actual_words = set(re.findall(r"\w+", actual.lower()))

        if not target_words:
            overlap = 0.0
        else:
            intersection = target_words.intersection(actual_words)
            overlap = len(intersection) / len(target_words)

        score_val = round(overlap * 70.0, 1)
        return ScoreResult(
            scorer_name=self.name,
            score=score_val,
            unit="percent",
            details={
                "exact_match": False,
                "token_overlap_ratio": round(overlap, 3),
            },
            reasoning=f"Token overlap ratio: {overlap:.1%} with reference answer",
        )


class LLMJudgeScorer(BaseScorer):
    """Uses a designated LLM runner as a judge to evaluate and score runner outputs based on a rubric."""

    DEFAULT_RUBRIC = (
        "Evaluate the response on a scale from 0 to 100 based on:\n"
        "1. Correctness & Accuracy (40%)\n"
        "2. Completeness & Code Quality (30%)\n"
        "3. Conciseness & Clarity (20%)\n"
        "4. Strict adherence to instructions (10%)\n"
    )

    def __init__(
        self,
        judge_runner: BaseRunner,
        rubric: str | None = None,
    ) -> None:
        self.judge_runner = judge_runner
        self.rubric = rubric or self.DEFAULT_RUBRIC

    @property
    def name(self) -> str:
        return f"llm_judge ({self.judge_runner.id})"

    async def score(
        self,
        task: str,
        result: RunnerResult,
        reference_answer: str | None = None,
        all_results: list[RunnerResult] | None = None,
    ) -> ScoreResult:
        if not result.success:
            return ScoreResult(
                scorer_name=self.name,
                score=0.0,
                unit="points",
                details={"judge_model": self.judge_runner.model},
                reasoning="Output failed execution, 0 points assigned",
            )

        ref_section = (
            f"\n\n[Optional Reference Answer]:\n{reference_answer}"
            if reference_answer
            else ""
        )

        judge_prompt = f"""You are an impartial, expert AI benchmark judge.
Your role is to evaluate an LLM's response to a given user task according to the scoring rubric.

[Evaluation Rubric]:
{self.rubric}

[User Task/Prompt]:
{task}
{ref_section}

[Candidate Response to Evaluate]:
{result.output}

Provide your evaluation strictly in valid JSON format with the following keys:
- "score": A numeric score from 0.0 to 100.0 (float).
- "strengths": A short list of strengths (strings).
- "weaknesses": A short list of weaknesses (strings).
- "critique": A 1-2 sentence overall summary of the evaluation.

Respond ONLY with valid JSON, without any markdown code fence wrappers if possible.
"""

        try:
            judge_res = await self.judge_runner.execute(
                task=judge_prompt,
                params={"temperature": 0.1, "max_tokens": 512},
            )

            if not judge_res.success:
                return ScoreResult(
                    scorer_name=self.name,
                    score=50.0,
                    unit="points",
                    details={"error": judge_res.error_message},
                    reasoning=f"Judge failed to evaluate: {judge_res.error_message}",
                )

            cleaned_output = judge_res.output.strip()
            # Strip markdown code fences if judge wrapped in ```json ... ```
            if cleaned_output.startswith("```"):
                cleaned_output = re.sub(r"^```(?:json)?\s*", "", cleaned_output)
                cleaned_output = re.sub(r"\s*```$", "", cleaned_output)

            parsed = json.loads(cleaned_output)
            score_val = float(parsed.get("score", 50.0))
            # Clamp score between 0.0 and 100.0
            score_val = max(0.0, min(100.0, score_val))

            return ScoreResult(
                scorer_name=self.name,
                score=score_val,
                unit="points",
                details={
                    "judge_runner": self.judge_runner.id,
                    "judge_model": self.judge_runner.model,
                    "strengths": parsed.get("strengths", []),
                    "weaknesses": parsed.get("weaknesses", []),
                    "critique": parsed.get("critique", ""),
                },
                reasoning=parsed.get("critique", f"Judge assigned {score_val}/100"),
            )

        except Exception as e:
            return ScoreResult(
                scorer_name=self.name,
                score=50.0,
                unit="points",
                details={"parse_error": str(e)},
                reasoning=f"Judge output parsing failed: {str(e)}",
            )

File: src/benchmark/scorers.ts
/** Pluggable scorers for benchmarking LLM outputs. */

import type { BaseRunner, RunnerResult } from '../runners/base.js';
import { errorMessage, round } from '../util.js';

/**
 * Evaluation score and metrics returned by a scorer.
 *
 * snake_case and in declaration order: this is serialized straight onto the wire.
 */
export interface ScoreResult {
  scorer_name: string;
  /** Normalized score (typically 0-100) or a raw metric value. */
  score: number;
  /** "points", "ms", "usd", "percent", "tokens/sec". */
  unit: string;
  details: Record<string, unknown>;
  reasoning: string | null;
}

function makeScore(init: {
  scorer_name: string;
  score: number;
  unit?: string;
  details?: Record<string, unknown>;
  reasoning?: string | null;
}): ScoreResult {
  return {
    scorer_name: init.scorer_name,
    score: init.score,
    unit: init.unit ?? 'points',
    details: init.details ?? {},
    reasoning: init.reasoning ?? null,
  };
}

/** Context passed to every scorer. */
export interface ScoreContext {
  task: string;
  result: RunnerResult;
  referenceAnswer?: string | null;
  allResults?: RunnerResult[] | null;
}

/** Abstract base class for benchmark scorers. */
export abstract class BaseScorer {
  abstract get name(): string;

  /** Compute an evaluation score for a runner's result. */
  abstract score(context: ScoreContext): Promise<ScoreResult>;
}

/** Scores runners on response latency and generation throughput. */
export class LatencyScorer extends BaseScorer {
  override get name(): string {
    return 'latency';
  }

  override async score({ result }: ScoreContext): Promise<ScoreResult> {
    if (!result.success) {
      return makeScore({
        scorer_name: this.name,
        score: 0,
        unit: 'ms',
        details: { latency_ms: result.latency_ms, tokens_per_second: 0 },
        reasoning: 'Execution failed',
      });
    }

    let tokensPerSec = 0;
    if (result.latency_ms > 0 && result.output_tokens > 0) {
      tokensPerSec = round(result.output_tokens / (result.latency_ms / 1000), 2);
    }

    return makeScore({
      scorer_name: this.name,
      score: result.latency_ms,
      unit: 'ms',
      details: {
        latency_ms: result.latency_ms,
        output_tokens: result.output_tokens,
        tokens_per_second: tokensPerSec,
      },
      reasoning: `Completed in ${result.latency_ms} ms (~${tokensPerSec} tokens/sec)`,
    });
  }
}

/** Scores runners on financial token cost in USD. */
export class CostScorer extends BaseScorer {
  override get name(): string {
    return 'cost';
  }

  override async score({ result }: ScoreContext): Promise<ScoreResult> {
    const totalTokens = result.input_tokens + result.output_tokens;

    return makeScore({
      scorer_name: this.name,
      score: result.estimated_cost_usd,
      unit: 'usd',
      details: {
        input_tokens: result.input_tokens,
        output_tokens: result.output_tokens,
        total_tokens: totalTokens,
        estimated_cost_usd: result.estimated_cost_usd,
      },
      reasoning: `Cost: $${result.estimated_cost_usd.toFixed(6)} across ${totalTokens} total tokens`,
    });
  }
}

/**
 * Scores output against a reference answer via exact match and token overlap.
 *
 * The word split uses `\p{L}\p{N}_` rather than JS's ASCII-only `\w` so the
 * overlap ratio matches Python's Unicode-aware `re.findall(r"\w+", ...)`.
 */
export class ExactMatchScorer extends BaseScorer {
  private static readonly WORD_PATTERN = /[\p{L}\p{N}_]+/gu;

  override get name(): string {
    return 'reference_match';
  }

  override async score({ result, referenceAnswer }: ScoreContext): Promise<ScoreResult> {
    if (!referenceAnswer) {
      return makeScore({
        scorer_name: this.name,
        score: 0,
        unit: 'percent',
        details: { evaluated: false },
        reasoning: 'No reference answer provided for matching',
      });
    }

    if (!result.success || !result.output) {
      return makeScore({
        scorer_name: this.name,
        score: 0,
        unit: 'percent',
        details: { exact_match: false, token_overlap: 0 },
        reasoning: 'Runner did not produce valid output',
      });
    }

    const target = referenceAnswer.trim();
    const actual = result.output.trim();

    if (actual === target) {
      return makeScore({
        scorer_name: this.name,
        score: 100,
        unit: 'percent',
        details: { exact_match: true, token_overlap: 1 },
        reasoning: 'Exact verbatim match with reference answer',
      });
    }

    if (actual.toLowerCase() === target.toLowerCase()) {
      return makeScore({
        scorer_name: this.name,
        score: 95,
        unit: 'percent',
        details: { exact_match: false, case_insensitive_match: true },
        reasoning: 'Case-insensitive match with reference answer',
      });
    }

    if (actual.toLowerCase().includes(target.toLowerCase())) {
      return makeScore({
        scorer_name: this.name,
        score: 80,
        unit: 'percent',
        details: { substring_match: true },
        reasoning: 'Reference answer is fully contained in output',
      });
    }

    const targetWords = new Set(target.toLowerCase().match(ExactMatchScorer.WORD_PATTERN) ?? []);
    const actualWords = new Set(actual.toLowerCase().match(ExactMatchScorer.WORD_PATTERN) ?? []);

    let overlap = 0;
    if (targetWords.size > 0) {
      let intersection = 0;
      for (const word of targetWords) if (actualWords.has(word)) intersection++;
      overlap = intersection / targetWords.size;
    }

    return makeScore({
      scorer_name: this.name,
      score: round(overlap * 70, 1),
      unit: 'percent',
      details: { exact_match: false, token_overlap_ratio: round(overlap, 3) },
      reasoning: `Token overlap ratio: ${(overlap * 100).toFixed(1)}% with reference answer`,
    });
  }
}

/** Uses a designated LLM runner as a judge to score outputs against a rubric. */
export class LLMJudgeScorer extends BaseScorer {
  static readonly DEFAULT_RUBRIC =
    'Evaluate the response on a scale from 0 to 100 based on:\n' +
    '1. Correctness & Accuracy (40%)\n' +
    '2. Completeness & Code Quality (30%)\n' +
    '3. Conciseness & Clarity (20%)\n' +
    '4. Strict adherence to instructions (10%)\n';

  readonly rubric: string;

  constructor(
    readonly judgeRunner: BaseRunner,
    rubric?: string | null,
  ) {
    super();
    this.rubric = rubric || LLMJudgeScorer.DEFAULT_RUBRIC;
  }

  override get name(): string {
    return `llm_judge (${this.judgeRunner.id})`;
  }

  override async score({ task, result, referenceAnswer }: ScoreContext): Promise<ScoreResult> {
    if (!result.success) {
      return makeScore({
        scorer_name: this.name,
        score: 0,
        unit: 'points',
        details: { judge_model: this.judgeRunner.model },
        reasoning: 'Output failed execution, 0 points assigned',
      });
    }

    const refSection = referenceAnswer ? `\n\n[Optional Reference Answer]:\n${referenceAnswer}` : '';

    const judgePrompt = `You are an impartial, expert AI benchmark judge.
Your role is to evaluate an LLM's response to a given user task according to the scoring rubric.

[Evaluation Rubric]:
${this.rubric}

[User Task/Prompt]:
${task}
${refSection}

[Candidate Response to Evaluate]:
${result.output}

Provide your evaluation strictly in valid JSON format with the following keys:
- "score": A numeric score from 0.0 to 100.0 (float).
- "strengths": A short list of strengths (strings).
- "weaknesses": A short list of weaknesses (strings).
- "critique": A 1-2 sentence overall summary of the evaluation.

Respond ONLY with valid JSON, without any markdown code fence wrappers if possible.
`;

    try {
      const judgeResult = await this.judgeRunner.execute(judgePrompt, { temperature: 0.1, max_tokens: 512 });

      if (!judgeResult.success) {
        return makeScore({
          scorer_name: this.name,
          score: 50,
          unit: 'points',
          details: { error: judgeResult.error_message },
          reasoning: `Judge failed to evaluate: ${judgeResult.error_message}`,
        });
      }

      let cleaned = judgeResult.output.trim();
      if (cleaned.startsWith('```')) {
        cleaned = cleaned.replace(/^```(?:json)?\s*/, '').replace(/\s*```$/, '');
      }

      const parsed = JSON.parse(cleaned) as Record<string, unknown>;

      const rawScore = 'score' in parsed ? Number(parsed['score']) : 50;
      if (!Number.isFinite(rawScore)) throw new Error(`could not convert score to float: ${String(parsed['score'])}`);
      const scoreValue = Math.max(0, Math.min(100, rawScore));

      const critique = 'critique' in parsed ? String(parsed['critique']) : `Judge assigned ${scoreValue}/100`;

      return makeScore({
        scorer_name: this.name,
        score: scoreValue,
        unit: 'points',
        details: {
          judge_runner: this.judgeRunner.id,
          judge_model: this.judgeRunner.model,
          strengths: parsed['strengths'] ?? [],
          weaknesses: parsed['weaknesses'] ?? [],
          critique: 'critique' in parsed ? parsed['critique'] : '',
        },
        reasoning: critique,
      });
    } catch (error) {
      return makeScore({
        scorer_name: this.name,
        score: 50,
        unit: 'points',
        details: { parse_error: errorMessage(error) },
        reasoning: `Judge output parsing failed: ${errorMessage(error)}`,
      });
    }
  }
}
